### PR TITLE
fix moving folders out of a cache jail

### LINKED ISSUE #3502: Duplicated files and folders when moving from one shared folder to another (can reproduce it)
### Steps to reproduce

through the web-ui:
1. UserA creates FolderA and FolderB
2. UserA shares FolderA and FolderB with UserB

on win7 desktop with client v2.2.4.2:
1. UserB creates the folder FOO in FolderA
2. UserB creates the folder BAR in FolderA/FOO
3. Wait for the desktop client  to sync these changes
4. UserB perform a cut-paste to move FOO directly from FolderA to FolderB (moving FOO on the desktop and then into FolderB does not make the issue to occur)
5. Wait for the desktop client to sync this change
6. UserB creates the folder POOP in FolderB/FOO/BAR/
7. Wait for the desktop client to sync this change

### Expected behaviour
Moving content from one share to another should work smoothly

### Actual behaviour
**The following inconsistency has appeared on the server: through the web-ui UserB can see the BAR folder inexplicably duplicated.**

### Other Probable side-effects
If FolderA and FolderB is shared with several other users working on them,  users keep getting the "Backup Detected" dialog asking wether to keep conflict files or to synchronize normally.

### Probable related issues or discussions
[https://help.nextcloud.com/t/duplicate-files-delete-not-possible/8869/7](https://help.nextcloud.com/t/duplicate-files-delete-not-possible/8869/7)
[https://github.com/nextcloud/server/issues/741](https://github.com/nextcloud/server/issues/741)

### Server configuration
**Operating system**:
Ubuntu 16.04

**Web server:**
Apache 2.4

**Database:**
Mysql 5.7

**PHP version:**
v5.6

**Nextcloud version:** (see Nextcloud admin page)
v10.0.3

**Updated from an older Nextcloud/ownCloud or fresh install:**
updated from 10.0.0 -> ... -> 10.0.3

**Where did you install Nextcloud from:**
official tarball

**Signing status:**
No errors have been found.

**List of activated apps:**
Enabled:
  - activity: 2.3.2
  - admin_audit: 1.0.0
  - comments: 1.0.0
  - dav: 1.0.1
  - federatedfilesharing: 1.0.1
  - federation: 1.0.1
  - files: 1.5.2
  - files_external: 1.0.2
  - files_pdfviewer: 0.8.1
  - files_sharing: 1.0.0
  - files_texteditor: 2.1
  - files_trashbin: 1.0.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 15.0.1
  - notifications: 0.3.0
  - password_policy: 1.0.0
  - provisioning_api: 1.0.0
  - serverinfo: 1.1.1
  - survey_client: 0.1.5
  - systemtags: 1.0.2
  - theming: 1.0.1
  - updatenotification: 1.0.1
  - workflowengine: 1.0.1
Disabled:
  - encryption
  - external
  - files_accesscontrol
  - files_automatedtagging
  - files_retention
  - files_versions
  - templateeditor
  - user_external
  - user_ldap
  - user_saml

**The content of config/config.php:**
```
{
    "system": {
        "instanceid": "ocq05beinini",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "aaa.bbb.net"
        ],
        "datadirectory": "\/var\/www\/nextcloud-data",
        "overwrite.cli.url": "http:\/\/aaa.bbb.net",
        "dbtype": "mysql",
        "version": "9.1.3.2",
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "loglevel": 0,
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "localhost",
            "port": 6379
        },
        "maintenance": false,
        "updater.release.channel": "stable"
    }
}
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** yes/no
No

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

### Client configuration
**Browser:**
Chrome 56

**Operating system:**
Windows 7

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes https://github.com/nextcloud/server/issues/3502

@plrunner @mrieser @bat553 please verify if this fixes the problem for you

### CODE REVIEW

**Comment by warnerbryce** [comment]
Hello @icewind1991 

How can i test this fix i don't have this path in my cloud : tests/lib/Files/Cache/Wrapper/CacheJailTest.php

I have the CacheJail.php in lib/private/Files/Cache/Wrapper/ 

Does this CacheJailTest.php is necessary ?

**Comment by LukasReschke** [comment]
> Does this CacheJailTest.php is necessary ?

This is only the unit test, it is not packaged within the final releases. So you don't have to replace that file, it's only for our automated tests relevant :)

**Review by nickvergessen — DISMISSED** [review]

**Comment by LukasReschke** [comment]
@icewind1991 Can we have a repair step here? If possible a live repair step to not block the upgrade and in multiple smaller steps.

**Review by LukasReschke — DISMISSED** [review]
This change works for me locally, but we need to have a way to repair older installations. But there's another desktop client bug … let me try to reproduce this and file to the client repo 🙈

**Comment by icewind1991** [comment]
note that this issue is completely reproducible without the client by moving files from the web ui

**Comment by bat553** [comment]
I have not yet been able to apply the fix but I wanted to report that this bug also happened when we create a shared folder.

**Comment by icewind1991** [comment]
@LukasReschke added repair step

**Comment by LukasReschke** [comment]
@nickvergessen Can you review the repair step? THX!

**Comment by nickvergessen** [comment]
@icewind1991 please write a unit test for the repair step.
Should be fairly easy?!

**Comment by icewind1991** [comment]
@nickvergessen done

**Comment by MorrisJobke** [comment]
I updated this here to do:

* move the repair step into a NC13 folder
* add a version number check to only run the repair step once

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍 (also the repair step properly cleans it up)

File: lib/private/Repair/NC13/RepairInvalidPaths.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+			} else {
+				$this->update($entry['fileid'], $calculatedPath);
+			}
+		}
+		return count($entries);
+	}
+
+	public function run(IOutput $output) {
+		$versionFromBeforeUpdate = $this->config->getSystemValue('version', '0.0.0');
+		if (version_compare($versionFromBeforeUpdate, '13.0.0.1', '<')) {
```
Comment (LukasReschke) [inline lib/private/Repair/NC13/RepairInvalidPaths.php]: Can we adjust this to either:

- Use an appconfig value
- Also check for 12.0.1 as we want to backport this one :)

Comment (MorrisJobke) [inline lib/private/Repair/NC13/RepairInvalidPaths.php]: > Use an appconfig value

This needs to be cleaned up later.



> Also check for 12.0.1 as we want to backport this one :)

Okay - if this is 12.0.1 then I will add it as well.

File: tests/lib/Repair/RepairInvalidPathsTest.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Test\Repair;
+
+use OC\Files\Cache\Cache;
+use OC\Files\Storage\Temporary;
+use OC\Repair\RepairInvalidPaths;
```
Comment (LukasReschke) [inline tests/lib/Repair/RepairInvalidPathsTest.php]: Test not passing, you moved that file 😉

**Review by LukasReschke — CHANGES_REQUESTED** [review]

**Comment by JiriMON** [comment]
Seems to work ok for moving folders/files out of a cache jail, but still seeing some occurrences of duplicate files as #3502. Tested on 12.0.0.29 - manually backported from changed files.

**Comment by jospoortvliet** [comment]
Note that it results in lost data :cry: :scream: 
I updated a file on my laptop but it doesn't sync to my desktop, even though both clients claim to be up to date.

**Comment by icewind1991** [comment]
@LukasReschke please re-review

**Comment by daita** [comment]
Tried to reproduce #3502 with linux client (2.3.2beta, build 13) and this branch. No can do. Works fine for me.

**Comment by MorrisJobke** [comment]
> Tried to reproduce #3502 with linux client (2.3.2beta, build 13) and this branch. No can do. Works fine for me.

Seeing as :+1: -> merging

**Comment by MorrisJobke** [comment]
@icewind1991 please create the backport PR ;)

**Pull request merged**